File: src/kea_fabric/api/state.py
"""In-memory mutable API state (discovery + perf tick). Layout uses JsonLayoutStore."""

from __future__ import annotations

from datetime import UTC, datetime

from kea_fabric.api.stub_mock_tables import MOCK_DISCOVERY_RECORD_COUNT

_discovery_paused: bool = False
_perf_tick: int = 0


def get_perf_tick() -> int:
    """Last perf tick (0 before first SSE `next_perf_tick`)."""
    return _perf_tick


def get_discovery_scan() -> dict[str, object]:
    state = "paused" if _discovery_paused else "running"
    return {
        "state": state,
        "updated_at": datetime.now(UTC).isoformat().replace("+00:00", "Z"),
        "record_count": MOCK_DISCOVERY_RECORD_COUNT,
    }


def set_discovery_paused(paused: bool) -> dict[str, object]:
    global _discovery_paused
    _discovery_paused = paused
    return get_discovery_scan()


def next_perf_tick() -> int:
    global _perf_tick
    _perf_tick += 1
    return _perf_tick


def reset_stub_state() -> None:
    """Clear mutable stub state (tests)."""
    global _discovery_paused, _perf_tick
    _discovery_paused = False
    _perf_tick = 0

File: apps/ui/src/mock/state.ts
import type { DashboardLayout } from "../lib/dashboard/types";

import { MOCK_DISCOVERY_RECORD_COUNT } from "./mockConstants";

export interface DiscoveryScanState {
  state: "idle" | "running" | "paused" | "failed";
  updated_at: string;
  record_count: number | null;
}

let discoveryScan: DiscoveryScanState = {
  state: "running",
  updated_at: "2026-04-22T10:00:00Z",
  record_count: MOCK_DISCOVERY_RECORD_COUNT,
};

let savedLayout: DashboardLayout | null = null;

/** Live mock for perf SSE — bumped periodically in dev. */
let perfTick = 0;

/** Current perf tick (last value after `nextPerfTick`; 0 before first SSE increment). */
export function getPerfTick(): number {
  return perfTick;
}

export function getDiscoveryScan(): DiscoveryScanState {
  return { ...discoveryScan };
}

export function setDiscoveryPaused(paused: boolean): DiscoveryScanState {
  discoveryScan = {
    ...discoveryScan,
    state: paused ? "paused" : "running",
    updated_at: new Date().toISOString(),
    record_count: discoveryScan.record_count,
  };
  return getDiscoveryScan();
}

export function getSavedLayout(): DashboardLayout | null {
  return savedLayout;
}

export function setSavedLayout(layout: DashboardLayout): void {
  savedLayout = layout;
}

export function nextPerfTick(): number {
  perfTick += 1;
  return perfTick;
}
